lib/listen/adapter/jruby.rb

```diff
@@ -0,0 +1,54 @@
+module Listen
+  module Adapter
+    # @see https://docs.oracle.com/javase/tutorial/essential/io/notification.html
+    class Jruby < Base
+      def self.usable?
+        RUBY_ENGINE == 'jruby'
+      end
+
+      private
+
+      def _configure(directory, &_callback)
+        require 'java'
+        java_import 'java.nio.file.FileSystems'
+        java_import 'java.nio.file.Paths'
+        java_import 'java.nio.file.StandardWatchEventKinds'
+
+        event_kinds = [
+          StandardWatchEventKinds::ENTRY_CREATE,
+          StandardWatchEventKinds::ENTRY_MODIFY,
+          StandardWatchEventKinds::ENTRY_DELETE
+        ]
+
+        @watcher ||= FileSystems.getDefault.newWatchService
+        @keys ||= {}
+        path = Paths.get(directory.to_s)
+        key = path.register(@watcher, *event_kinds)
+        @keys[key] = path
+      end
+
+      def _run
+        loop do
+          key = @watcher.take
+          dir = @keys[key]
+          next if dir.nil?
+          key.pollEvents.each do |event|
+            kind = event.kind
+            next if kind == StandardWatchEventKinds::OVERFLOW
+            name = event.context
+            child = dir.resolve(name)
+            pathname = Pathname.new(child.to_s).dirname
+            _queue_change(:dir, pathname, '.', recursive: true)
+            valid = key.reset
```

Comment:
Not a ruby dev, but:
- Should reset be called after this loop, not in it.
- Should reset be called regardless of the result of dir.nil?

Reply:
> * Should reset be called after this loop, not in it.

You're quite right - I got lost in all the braces in the Java example. I've addressed this in a fix-up commit.

> * Should reset be called regardless of the result of dir.nil?

It doesn't appear to be the case in the Java example, but what you say makes sense. I think in the Java example, they're probably assuming it can never happen, because every `path` that's registered will have an entry in the `keys` `Map`. The same thing is probably true of the current implementation, but I can see a case for doing as you suggest, so I've made the change in another fix-up commit.